src/language-js/embed.js

```diff
@@ -32,18 +65,95 @@ function embed(path, print, textToDoc, options) {
-        const rawQuasis = node.quasis.map(q => q.value.raw);
-        let placeholderID = 0;
-        const text = rawQuasis.reduce((prevVal, currVal, idx) => {
-          return idx == 0
-            ? currVal
-            : prevVal +
-                "@prettier-placeholder-" +
-                placeholderID++ +
-                "-id" +
-                currVal;
+        const text = node.quasis.reduce((prevVal, current, index) => {
+          const raw = current.value.raw;
+          return index === 0
+            ? raw
+            : prevVal + cssPlaceholder.get(index - 1) + raw;
         }, "");
-        const doc = textToDoc(text, { parser: "css" });
+
+        // css parser can't handle the following css
+        // ```css
+        // div {
+        //   css-placeholder;
+        // }
+        // ```
+        // so we fake it into
+        // ```css
+        // div {
+        //   prop-placeholder: css-placeholder;
+        // }
+        // ```
+        // and will restore back after parse
+
+        const pieces = cssPlaceholder.parse(text);
+        const texts = placeholderPiecesToStringArray(pieces);
+
+        pieces.forEach(({ isPlaceholder, placeholder }, index) => {
+          if (!isPlaceholder) {
+            return;
+          }
+
+          let text = placeholder;
+          let before = texts.slice(0, index).join();
+          let after = texts.slice(index + 1).join();
+
+          // move identity character
+          const leadingIdentity = findCSSIdentity(before, true);
+          const tailingIdentity = findCSSIdentity(after);
+          if (leadingIdentity) {
+            texts[index - 1] = texts[index - 1].slice(
+              0,
+              -leadingIdentity.length
+            );
+            text = leadingIdentity + text;
+          }
+          if (tailingIdentity) {
+            texts[index + 1] = texts[index + 1].slice(tailingIdentity.length);
+            text += tailingIdentity;
+          }
+
+          // check orphan placeholder
+          after = cleanCSS(
+            texts
+              .slice(index + 1)
+              .filter(text => !cssPlaceholder.isPlaceholder(text))
+              .join("")
+          );
+          const endsWithLineBreak = /^\s*\n/.test(after);
+          after = after.trim();
+
+          const needExtraSemicolon =
+            !after ||
+            endsWithLineBreak ||
+            after[0] !== ";" ||
+            cssPlaceholder.isPlaceholder(
+              texts.slice(index + 1).filter(text => text.trim())[0] || ""
+            );
+
+          before = cleanCSS(texts.slice(0, index).join("")).trim();
+
+          if (
+            (!after ||
+              endsWithLineBreak ||
+              after[0] === ";" ||
+              after[0] === "}") &&
+            (!before ||
+              before.slice(-1) === ";" ||
+              before.slice(-1) === "{" ||
+              before.slice(-1) === "}")
+          ) {
```

Comment:
Move this stuff under function with good names like `isEndOfDeclaration`/`isEndOfRule`/etc, because it is hard to understand what you mean